src/utils/AddressUtils.ts

```diff
@@ -68,3 +70,187 @@ export function isValidEvmAddress(address: string): boolean {
+  toHexString(): string {
+    return this.toBytes32();
+  }
+
+  // Checks if this address can be coerced into a bytes20 evm address. Returns true if it is possible and false otherwise.
+  isValidEvmAddress(): boolean {
+    try {
+      this.toAddress();
+      return true;
+    } catch {
+      return false;
+    }
+  }
+
+  // Checks if this address can be coerced into a valid Solana address. Return true if possible and false otherwise.
+  isValidSvmAddress(): boolean {
+    return isSvmAddress(this.toBase58());
+  }
+
+  // Checks if the address is valid on the given chain ID.
+  isValidOn(chainId: number): boolean {
+    if (chainIsEvm(chainId)) {
+      return this.isValidEvmAddress();
+    }
+    return this.isValidSvmAddress();
+  }
+
+  // Checks if the object is an address by looking at whether it has an Address constructor.
+  static isAddress(obj: unknown): boolean {
+    return obj instanceof this;
+  }
+
+  // Converts the input address to a 32-byte hex data string.
+  toString(): string {
+    return this.toHexString();
+  }
+
+  // Checks if the address is the zero address.
+  isZeroAddress(): boolean {
+    return utils.stripZeros(this.rawAddress).length === 0;
+  }
+
+  // Checks if the other address is equivalent to this address.
+  eq(other: Address): boolean {
+    return this.toString() === other.toString();
+  }
+
+  // Compares Addresses by first converting them to BigNumbers.
+  compare(otherAddress: Address): 1 | -1 | 0 {
+    // Convert address strings to BigNumbers and then sort numerical value of the BigNumber, which sorts the addresses
+    // effectively by their hex value.
+    const bnAddressA = this.toBigNumber();
+    const bnAddressB = otherAddress.toBigNumber();
+    if (bnAddressA.gt(bnAddressB)) {
+      return 1;
+    } else if (bnAddressA.lt(bnAddressB)) {
+      return -1;
+    } else {
+      return 0;
+    }
+  }
+}
+
+// Subclass of address which strictly deals with 20-byte addresses. These addresses are guaranteed to be valid EVM addresses, so `toAddress` will always succeed.
+export class EvmAddress extends Address {
+  // On construction, validate that the address can indeed be coerced into an EVM address. Throw immediately if it cannot.
+  constructor(rawAddress: Uint8Array) {
+    super(rawAddress);
+    const hexString = utils.hexlify(rawAddress);
+    if (!this.isValidEvmAddress()) {
+      throw new Error(`${hexString} is not a valid EVM address`);
+    }
+  }
+
+  // Override `toAddress` to return the 20-byte representation address.
+  override toAddress(): string {
+    return this.toEvmAddress();
+  }
+
+  // Constructs a new EvmAddress type.
+  static from(hexString: string): EvmAddress {
+    return new this(utils.arrayify(hexString));
+  }
+}
+
+// Subclass of address which strictly deals SVM addresses. These addresses are guaranteed to be valid SVM addresses, so `toBase58` will always produce a valid Solana address.
+export class SvmAddress extends Address {
```

Comment:
How do you think we should handle translation from bytes20 to bytes32. This is important when we see a PoolRebalanceRoute/CrossChainContractsSet event from the HubPool which emits `address`'s that we'll have to map to Solana addresses. The [SolanaAdapter](https://github.com/across-protocol/contracts/blob/a6bd1bdc948b0eb0075cc0e753a26d2ff70acd6e/contracts/chain-adapters/Solana_Adapter.sol#L45) stores hardcoded btyes32 addresses of USDC (the only Solana token we expect to support for now) and the Solana_SpokePool, so I wonder if we need to hardcode those addresses in here as well.

Reply:
Yeah; I think we need to read this translation onchain via config (i.e. for each PoolRebalanceRoute token, we query some other contract for the SVM representation of it). Ideally we can do this via a multicall, rather than having to index events.